google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/Timestamp.java

```diff
@@ -176,14 +165,64 @@ public static Timestamp fromProto(com.google.protobuf.Timestamp proto) {
     return com.google.protobuf.Timestamp.newBuilder().setSeconds(seconds).setNanos(nanos).build();
   }
 
+  private static final int[] POWERS_OF_10 = {
+    1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000
+  };
+
   /**
    * Creates a Timestamp instance from the given string. String is in the RFC 3339 format without
    * the timezone offset (always ends in "Z").
    */
   public static Timestamp parseTimestamp(String timestamp) {
-    TemporalAccessor temporalAccessor = timestampParser.parse(timestamp);
-    Instant instant = Instant.from(temporalAccessor);
-    return ofTimeSecondsAndNanos(instant.getEpochSecond(), instant.getNano());
+    Preconditions.checkNotNull(timestamp);
+    final String invalidTimestamp = "Invalid timestamp: " + timestamp;
+    Preconditions.checkArgument(
+        timestamp.length() >= 19 && timestamp.length() <= 30, invalidTimestamp);
+    Preconditions.checkArgument(timestamp.charAt(4) == '-', invalidTimestamp);
+    Preconditions.checkArgument(timestamp.charAt(7) == '-', invalidTimestamp);
+    Preconditions.checkArgument(
+        timestamp.charAt(10) == 'T' || timestamp.charAt(10) == 't', invalidTimestamp);
+    Preconditions.checkArgument(
+        (timestamp.length() == 19 && (timestamp.charAt(18) != 'Z' || timestamp.charAt(18) == 'z'))
+            || (timestamp.length() == 20
+                && (timestamp.charAt(19) == 'Z' || timestamp.charAt(19) == 'z'))
+            || (timestamp.charAt(19) == '.'
+                && timestamp.length() > 20
+                && (timestamp.charAt(20) != 'Z' || timestamp.charAt(20) == 'z')),
+        invalidTimestamp);
+    try {
+      int year = Integer.parseInt(timestamp.substring(0, 4));
+      int month = Integer.parseInt(timestamp.substring(5, 7));
+      int day = Integer.parseInt(timestamp.substring(8, 10));
+      int hour = Integer.parseInt(timestamp.substring(11, 13));
+      int minute = Integer.parseInt(timestamp.substring(14, 16));
+      int second = Integer.parseInt(timestamp.substring(17, 19));
+      int fraction = 0;
+      if (timestamp.length() > 20) {
+        if (timestamp.charAt(19) == '.') {
+          int endIndex;
+          if (timestamp.charAt(timestamp.length() - 1) == 'Z'
+              || timestamp.charAt(timestamp.length() - 1) == 'z') {
+            endIndex = timestamp.length() - 1;
+          } else {
+            endIndex = timestamp.length();
+          }
+          if (endIndex - 20 > 9) {
+            throw new IllegalArgumentException(invalidTimestamp);
+          }
+          fraction = Integer.parseInt(timestamp.substring(20, endIndex));
+          // Adjust the result to nanoseconds if the input length is less than 9 digits (9 -
+          // (endIndex
+          // - 20)).
```

Comment:
Please have the mathematical formula on a single line, if possible.  `(9 - (endIndex - 20))`.